esp/clients/ws_dfsclient/ws_dfsclient.cpp

```diff
@@ -0,0 +1,702 @@
+class CDFSFile : public CSimpleInterfaceOf<IDFSFile>
+{
+    Linked<IPropertyTree> commonMeta; // e.g. share info between IFDSFiles, e.g. common plane info between subfiles
+    Linked<IPropertyTree> fileMeta;
+    unsigned __int64 lockId;
+    std::vector<Owned<IDFSFile>> subFiles;
+    StringAttr service;
+    unsigned timeoutSecs;
+    Linked<IUserDescriptor> userDesc;
+
+public:
+    CDFSFile(IPropertyTree *_commonMeta, IPropertyTree *_fileMeta, const char *_service, unsigned _timeoutSecs, IUserDescriptor *_userDesc)
+        : commonMeta(_commonMeta), fileMeta(_fileMeta), service(_service), timeoutSecs(_timeoutSecs), userDesc(_userDesc)
+    {
+        lockId = fileMeta->getPropInt64("@lockId");
+        if (fileMeta->getPropBool("@isSuper"))
+        {
+            Owned<IPropertyTreeIterator> iter = fileMeta->getElements("FileMeta");
+            ForEach(*iter)
+            {
+                IPropertyTree &subMeta = iter->query();
+                subFiles.push_back(createDFSFile(commonMeta, &subMeta, service, timeoutSecs, userDesc));
+            }
+        }
+    }
+    virtual IPropertyTree *queryFileMeta() const override
+    {
+        return fileMeta;
+    }
+    virtual IPropertyTree *queryCommonMeta() const override
+    {
+        return commonMeta;
+    }
+    virtual unsigned __int64 getLockId() const override
+    {
+        return lockId;
+    }
+    virtual unsigned numSubFiles() const override
+    {
+        return (unsigned)subFiles.size();
+    }
+    virtual IDFSFile *getSubFile(unsigned idx) const override
+    {
+        return LINK(subFiles[idx]);
+    }
+    virtual const char *queryService() const override
+    {
+        return service;
+    }
+    virtual IUserDescriptor *queryUserDescriptor() const override
+    {
+        return userDesc.getLink();
+    }
+    virtual unsigned queryTimeoutSecs() const override
+    {
+        return timeoutSecs;
+    }
+};
+
+static IDFSFile *createDFSFile(IPropertyTree *commonMeta, IPropertyTree *fileMeta, const char *service, unsigned timeoutSecs, IUserDescriptor *userDesc)
+{
+    return new CDFSFile(commonMeta, fileMeta, service, timeoutSecs, userDesc);
+}
+
+IClientWsDfs *getDfsClient(const char *serviceUrl, IUserDescriptor *userDesc)
+{
+    // JCSMORE - can I reuse these, are they thread safe (AFishbeck?)
+
+    VStringBuffer dfsUrl("%s/WsDfs", serviceUrl);
+    Owned<IClientWsDfs> dfsClient = createWsDfsClient();
+    dfsClient->addServiceUrl(dfsUrl);
+    StringBuffer user, token;
+    userDesc->getUserName(user),
+    userDesc->getPassword(token);
+    dfsClient->setUsernameToken(user, token, "");
+    return dfsClient.getClear();
+}
+
+static CriticalSection serviceLeaseMapCS;
+static std::unordered_map<std::string, unsigned __int64> serviceLeaseMap;
+unsigned __int64 ensureClientLease(const char *service, IUserDescriptor *userDesc)
+{
+    CriticalBlock block(serviceLeaseMapCS);
+    auto r = serviceLeaseMap.find(service);
+    if (r != serviceLeaseMap.end())
```

Comment:
Nothing ever removes items from serviceLeaseMap.  Is that intentional?

Reply:
this is really placeholder code/work in progress, related to locking, and will likely change, but I envisage these will remain indefinitely, unless it tries to contact a previous  esp and fails, and must remove and restablish a new leaseid with a different esp url.